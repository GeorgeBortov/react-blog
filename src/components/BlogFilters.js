import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export class BlogFilters extends React.Component {
    state = {
        
    };
    render() {
        return (
            <div className="page-header">
                <div className="content-container">
                    <form>
                        <div className="input-group">
                            <div className="input-group__item">
                                <input
                                    type="text"
                                    className="text-input"
                                    placeholder="Search posts by title"
                                    // value={this.props.filters.text}
                                    // onChange={this.onTextChange}
                                />
                            </div>
                            <div className="input-group__item">
                                <select
                                    className="select"
                                    // value={this.props.filters.sortBy}
                                    // onChange={this.onSortChange}
                                >
                                    <option value="date">Sort by date</option>
                                    <option value="cur_user">Select only my posts</option>
                                </select>
                            </div>
                        </div>
                    </form>
                    <Link className="button" to="/create">Add Post</Link>
                </div>
            </div>
        )
    }
}



export default connect(undefined, undefined)(BlogFilters);