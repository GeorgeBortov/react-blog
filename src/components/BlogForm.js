import React from 'react';
import { connect } from 'react-redux';
import JoditEditor from "jodit-react";
import moment from 'moment';

export class BlogForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: props.post ? props.post.title : '',
            body: props.post ? props.post.body : '',
            createdAt: moment(),
            authorName: props.userName ? props.userName : (props.post.authorName ? props.post.authorName : '' ),
            error: ''
        }
    }
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
    };
    onAuthorChange = (e) => {
        const authorName = e.target.value;
        this.setState(() => ({ authorName }));
    };
    onBodyChange = (value) => {
        const body = value;
        this.setState(() => ({ body }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.title || !this.state.body) {
            // 'Please provide description and amount'
            this.setState(() => ({ error: 'Please provide title and post body' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                title: this.state.title,
                body: this.state.body,
                createdAt: this.state.createdAt.valueOf(),
                authorName: this.state.authorName
            });
        }
    };
    config = {
		readonly: false // all options from https://xdsoft.net/jodit/doc/
	}
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Post title"
                    autoFocus
                    className="text-input form--title"
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <JoditEditor
                    editorRef={this.setRef}
                    value={this.state.body}
                    config={this.config}
                    onChange={this.onBodyChange}
                />
                {
                    !this.props.userName &&
                    <input
                        type="text"
                        placeholder="Author"
                        className="text-input form--author"
                        value={this.state.authorName}
                        onChange={this.onAuthorChange}
                    />
                }
                
                <div>
                    <button className="button">Save Post</button>
                </div>
            </form>
        )
    }
}

const mapStateToProps = (state) => ({
    userName: state.auth.name || ''
});

export default connect(mapStateToProps, undefined)(BlogForm);