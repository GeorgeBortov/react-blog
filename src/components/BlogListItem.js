import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import moment from 'moment';
import trimText from '../selectors/trim-text';
export const BlogListItem = ({ id, title, trimText, authorName, createdAt, uid, authorID, }) => (
    <div className="list-item">
        <div>
            <h3 className="list-item__title">{title}</h3>
            <div className="list-item__text">
                {trimText}
            </div>
        </div>
        <div className="list-item__data">
            <span className="list-item__author">Posted by {authorName !== '' ? authorName : 'Anonymous'}</span>
            <span className="list-item__time">{moment(createdAt).format('MMMM Do YYYY')} at {moment(createdAt).format('h:mm')}</span>
        </div>
        <div className="list-item__buttons">
            <Link className="button" to={`/read/${id}`}>Read</Link>
            {   
                uid === authorID &&
                    <Link className="button" to={`/edit/${id}`}>Edit</Link>
                
            }
            
        </div>
    </div>
    
);

const mapStateToProps = (state, ownProps) => ({
    uid: state.auth.uid,
    trimText: trimText(ownProps.body, 500)
});

const ConnectedBlogListItem = connect(mapStateToProps)(BlogListItem);

export default ConnectedBlogListItem;