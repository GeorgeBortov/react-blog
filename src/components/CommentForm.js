import React from 'react';
import { connect } from 'react-redux';
import JoditEditor from "jodit-react";
import moment from 'moment';

export class BlogForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            rate: props.post ? props.post.body : '',
            body: props.post ? props.post.body : '',
            createdAt: moment(),
            authorName: props.userName,
            error: ''
        }
        this.baseState = this.state
    }
    handleRateChange = (e) => {
        const rate = e.target.value;
        this.setState(() => ({ rate })); 
    };
    onAuthorChange = (e) => {
        const authorName = e.target.value;
        this.setState(() => ({ authorName }));
    };
    onBodyChange = (value) => {
        const body = value;
        this.setState(() => ({ body }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.authorName || !this.state.body || !this.state.rate) {
            // 'Please provide description and amount'
            this.setState(() => ({ error: 'Please fill in all fields and rate the post' }));
        } else {
            e.target.reset();
            this.props.onSubmit({
                rate: this.state.rate,
                body: this.state.body,
                createdAt: this.state.createdAt.valueOf(),
                authorName: this.state.authorName
            });
            this.setState(this.baseState);    
        }
    };
    config = {
		readonly: false // all options from https://xdsoft.net/jodit/doc/
	}
    render() {
        return (
            <div className="comment-form">
                
                <h4>Write a commpent</h4>
                
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <span className="rating">
                        <input type="radio" id="star5" name="rating" value="5" onBlur={this.handleRateChange}  /><label htmlFor="star5" title="Rocks!">5 stars</label>
                        <input type="radio" id="star4" name="rating" value="4" onBlur={this.handleRateChange}  /><label htmlFor="star4" title="Pretty good">4 stars</label>
                        <input type="radio" id="star3" name="rating" value="3" onBlur={this.handleRateChange}  /><label htmlFor="star3" title="Meh">3 stars</label>
                        <input type="radio" id="star2" name="rating" value="2" onBlur={this.handleRateChange}  /><label htmlFor="star2" title="Kinda bad">2 stars</label>
                        <input type="radio" id="star1" name="rating" value="1" onBlur={this.handleRateChange}  /><label htmlFor="star1" title="Sucks big time">1 star</label>
                        <span className="txt">Please rate:</span>
                    </span>
                    {
                        !this.props.userName &&
                        <input
                            type="text"
                            placeholder="Your name"
                            className="text-input form--author"
                            value={this.state.authorName}
                            onChange={this.onAuthorChange}
                        />
                    }
                    <JoditEditor
                        editorRef={this.setRef}
                        value={this.state.body}
                        config={this.config}
                        onChange={this.onBodyChange}
                    /> 
                    <div>
                        <button className="button">Publish</button>
                    </div>
                </form>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    userName: state.auth.name || state.auth.email
});

export default connect(mapStateToProps, undefined)(BlogForm);