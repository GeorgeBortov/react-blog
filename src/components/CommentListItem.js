import React from 'react';
import moment from 'moment';
import { Markup } from 'interweave';
export const CommentListItem = ({ authorName, body, createdAt, rate }) => (
    <div className="list-item">
        <div>
            <div className="star-ratings-wrap">
                <div className="star-ratings">
                    <span style={{width: rate*100/5 + '%'}} className="star-ratings-sprite-rating"></span>
                </div>
                <span className="score">Rating: {rate}/5</span>
            </div>
            <div className="list-item__text">
                <Markup content={body} />
            </div>
        </div>
        <div className="list-item__data">
            <span className="list-item__author">Posted by {authorName !== '' ? authorName : 'Anonymous'}</span>
            <span className="list-item__time">{moment(createdAt).format('MMMM Do YYYY')} at {moment(createdAt).format('h:mm')}</span>
            
        </div>
    </div>
    
);


export default CommentListItem;