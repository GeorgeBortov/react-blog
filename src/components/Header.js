import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout, isAuthenticated }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard">
                    <h1>Blog</h1>
                </Link>
                {isAuthenticated ? 
                    <button id='logout' className="button button--link" onClick={startLogout}>Logout</button> 
                    : 
                    <Link id='login' className="button button--link" to="/login">
                        Login
                    </Link>
                }
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
});


export default connect(mapStateToProps, mapDispatchToProps)(Header);