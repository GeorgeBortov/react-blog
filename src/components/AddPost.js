import React from 'react';
import { connect } from 'react-redux';
import BlogForm from './BlogForm';
import { startAddPost } from '../actions/posts';

export class AddPost extends React.Component {
    onSubmit = (post) => {
        this.props.startAddPost(post);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add post</h1>
                    </div>
                </div>
                <div className="content-container">
                    <BlogForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddPost: (post) => dispatch(startAddPost(post))
});

export default connect(undefined, mapDispatchToProps)(AddPost);