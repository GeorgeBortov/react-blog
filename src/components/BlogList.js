import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';


export const BlogList = (props) => {
    return (
        <div className="content-container">
            <div className="list-body">
                {   
                    props.posts.length === 0 ? (
                        <div className="list-item list-item--message">
                            <span>No posts</span>
                        </div>
                    ) : (
                        props.posts.map((post) => {
                            return <BlogListItem key={post.id} {...post} />
                        })
                    )
                }
            </div>
        </div>
    )
};

const mapStateToProps = (state) => ({
    posts: state.posts
});

const ConnectedBlogList = connect(mapStateToProps)(BlogList);

export default ConnectedBlogList;