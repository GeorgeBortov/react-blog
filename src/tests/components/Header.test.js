import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

test('should show button login if user is no auth', () => {
    const wrapper = shallow(<Header startLogout={() => {}} isAuthenticated={false} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#login').prop('to')).toBe('/login');
});
test('should show button logout if user is auth', () => {
    const wrapper = shallow(<Header startLogout={() => {}} isAuthenticated={true} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#logout').length).toBeGreaterThan(0);
});

test('should call startLogout on button click', () => {
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout={startLogout} isAuthenticated={true}/>);
    wrapper.find('#logout').simulate('click');
    expect(startLogout).toHaveBeenCalled();
});