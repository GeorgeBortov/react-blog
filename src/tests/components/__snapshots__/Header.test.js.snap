// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show button login if user is no auth 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Blog
        </h1>
      </Link>
      <Link
        className="button button--link"
        id="login"
        replace={false}
        to="/login"
      >
        Login
      </Link>
    </div>
  </div>
</header>
`;

exports[`should show button logout if user is auth 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Blog
        </h1>
      </Link>
      <button
        className="button button--link"
        id="logout"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
